File: DifferentWaysToAddParentheses.py
"""
241. Different Ways to Add Parentheses
Given a string of numbers and operators, return all possible results from computing all the different possible ways to group numbers and operators. The valid operators are +, - and *.


Example 1
Input: "2-1-1".

((2-1)-1) = 0
(2-(1-1)) = 2
Output: [0, 2]


Example 2
Input: "2*3-4*5"

(2*(3-(4*5))) = -34
((2*3)-(4*5)) = -14
((2*(3-4))*5) = -10
(2*((3-4)*5)) = -10
(((2*3)-4)*5) = 10
Output: [-34, -14, -10, -10, 10]
"""
# O(n * 4 ^ n / n ^ (3 / 2))
# O(n * 4 ^ n / n ^ (3 / 2))
# Catalan Number: h(n) = h(n - 1) * (4 * n - 2) / (n + 1)
class Solution(object):
    def diffWaysToCompute(self, input):
        """
        :type input: str
        :rtype: List[int]
        """
        if not input:
            return []
        l, tem = [], ""
        for i in range(len(input)):
            if input[i].isdigit():
                tem += input[i]
            else:
                l.append(int(tem))
                l.append(input[i])
                tem = ""
        l.append(int(tem))
        return self.recur(l)
        
    def recur(self, l):
        if len(l) == 1:
            return l
        res = []
        for i in range(1, len(l), 2):
            left = self.recur(l[0 : i])
            right = self.recur(l[i + 1 :])
            for j in left:
                for k in right:
                    res.append(self.opt(j, l[i], k)) 
        return res
    
    def opt(self, num1, operator, num2):
        if operator == "+":
            return num1 + num2
        if operator == "-":
            return num1 - num2
        return num1 * num2


class Solution2(object):
    def diffWaysToCompute(self, input):
        """
        :type input: str
        :rtype: List[int]
        """
        stackNum, stackOpt, tem = [], [], ""
        for i in input:
            if i in "+-*":
                stackNum.append(int(tem))
                stackOpt.append(i)
                tem = ""
            else:
                tem += i
        if tem:
            stackNum.append(int(tem))
        return self.recur(stackNum, stackOpt)
        
    def recur(self, stackNum, stackOpt):
        if not stackOpt:
            return stackNum
        res = []
        for i in range(len(stackOpt)):
            left = self.recur(stackNum[0:i + 1], stackOpt[0 : i])
            right = self.recur(stackNum[i + 1:], stackOpt[i + 1:])
            for l in left:
                for r in right:
                    res.append(self.opt(l, stackOpt[i], r))
        return res

    def opt(self, num1, opter, num2):
        if opter == "+":
            return num1 + num2
        if opter == "-":
            return num1 - num2
        return num1 * num2
    
if __name__ == "__main__":
    input = "1"
    for i in range(8):
        print(len(Solution().diffWaysToCompute(input)))
        input += "+1"

File: Java/DifferentWaysToAddParentheses.java
import java.util.*;
public class DifferentWaysToAddParentheses {
    public List<Integer> diffWaysToCompute(String input) {
        if(input.length() == 0) {return null;}
        List<Integer> nums = new ArrayList<Integer>();
        List<Character> opter = new ArrayList<Character>();
        int tem = 0;
        for(int i = 0; i < input.length(); i++) {
            if(input.charAt(i) >= '0' && input.charAt(i) <= '9') {
                tem = tem * 10 + input.charAt(i) - '0';
            } else {
                nums.add(tem);
                opter.add(input.charAt(i));
                tem = 0;
            }
        }
        nums.add(tem);
        return recur(nums, opter);
    }
    
    private List<Integer> recur(List<Integer> nums, List<Character> opter) {
        if(nums.size() <= 1) {return nums;}
        List<Integer> res = new ArrayList<Integer>();
        List<Integer> left;
        List<Integer> right;
        for(int i = 0; i < opter.size(); i++) {
            left = new ArrayList<Integer>(recur(nums.subList(0, i + 1), opter.subList(0, i)));
            right = new ArrayList<Integer>(recur(nums.subList(i + 1, nums.size()), opter.subList(i + 1, opter.size())));
            for(int l : left) {
                for(int r : right) {
                    res.add(calculate(l, r, opter.get(i)));
                }
            }
        }
        return res;
    }
    
    private int calculate(int l, int r, Character opter) {
        if(opter == '+') {return l + r;}
        if(opter == '-') {return l - r;}
        return l * r;
    }

}
